### PR TITLE
use BufferedIndexInput

### PR DESCRIPTION
**Pull request by rnewson**
This replaces my dodgy logic in FDBIndexInput for great good

### CODE REVIEW

**Comment by rnewson** [comment]
I really want much better test coverage for all the edge cases, so I'm looking into that.

File: src/test/java/com/cloudant/fdblucene/FDBIndexInputTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    return result;
+  }
+
+  private final long off;
+  private final long length;
+  private final int pageCount;
+  private final int pageSize;
+
+  public FDBIndexInputTest(long off, long length, int pageCount, int pageSize) {
```
Comment (ksnavely) [inline src/test/java/com/cloudant/fdblucene/FDBIndexInputTest.java]: Use `final` for method args where possible.

File: src/main/java/com/cloudant/fdblucene/FDBIndexInput.java [hunk 2]
```diff
@@ -1,146 +1,130 @@
...
+        page = loadPage(pageNumber);
+        lastPageNumber = pageNumber;
+      }
+      final int toCopy = Math.min(page.length - FDBUtil.posToOffset(pos, pageSize), readLength);
+      System.arraycopy(page, FDBUtil.posToOffset(pos, pageSize), b, offset + copied, toCopy);
+      pos += toCopy;
+      copied += toCopy;
+      readLength -= toCopy;
     }
+    assert readLength == 0;
```
Comment (tonysun83) [inline src/main/java/com/cloudant/fdblucene/FDBIndexInput.java]: does this ever happen?

File: src/main/java/com/cloudant/fdblucene/FDBIndexInput.java [hunk 3]
```diff
@@ -1,146 +1,130 @@
...
 
-    @Override
-    public byte readByte() throws IOException {
-        if (pointer == length) {
-            throw new EOFException("Attempt to read past end of file");
-        }
-
-        loadPageIfNull();
-        byte result = page[FDBUtil.posToOffset(offset + pointer, pageSize)];
-        seek(this.pointer + 1);
```
Comment (tonysun83) [inline src/main/java/com/cloudant/fdblucene/FDBIndexInput.java]: so the biggest difference from switching to BufferedIndexInput is that you are no longer utilizing `seek`. Was this the incorrect logic you were referring to?

File: src/main/java/com/cloudant/fdblucene/FDBIndexInput.java (hunk 2 above)
Comment (rnewson) [reply src/main/java/com/cloudant/fdblucene/FDBIndexInput.java]: it should never happen, that's why it's an assert. :)

Comment (rnewson) [reply src/main/java/com/cloudant/fdblucene/FDBIndexInput.java]: I modelled this after SimpleFSIndexInput, I even followed variable name conventions, as a way to avoid fence post errors.

File: src/main/java/com/cloudant/fdblucene/FDBIndexInput.java (hunk 3 above)
Comment (rnewson) [reply src/main/java/com/cloudant/fdblucene/FDBIndexInput.java]: no, I believe the incorrect logic is in the offset/length twiddling at some point (haven't identified where exactly). for page to be null after calling loadPageIfNull we'd need to have asked for a page beyond the end of the file, which obviously shouldn't happen. This new version feels safer by design.

File: src/test/java/com/cloudant/fdblucene/FDBIndexInputTest.java (hunk 1 above)
Comment (rnewson) [reply src/test/java/com/cloudant/fdblucene/FDBIndexInputTest.java]: done

**Comment by tonysun83** [comment]
Currently running the highrise workload  with this branch and these popped up:

```
2020-02-13 08:58:39 ERROR Search:301 - Catching
java.lang.IllegalArgumentException: cannot change field "iam_compatible" from index options=DOCS to inconsistent index options=DOCS_AND_FREQS_AND_POSITIONS
	at org.apache.lucene.index.FieldInfo.setIndexOptions(FieldInfo.java:258) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:792) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:442) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:406) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:250) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:495) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1594) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1586) ~[lucene-core-8.3.0.jar:8.3.0 2aa586909b911e66e1d8863aa89f173d69f86cd2 - ishan - 2019-10-25 23:10:03]
	at com.cloudant.search3.FDBDirectorySearchHandler.updateDocument(FDBDirectorySearchHandler.java:190) ~[classes/:?]
	at com.cloudant.search3.Search.lambda$updateDocument$4(Search.java:240) ~[classes/:?]
	at com.cloudant.search3.Search.execute(Search.java:294) [classes/:?]
	at com.cloudant.search3.Search.updateDocument(Search.java:237) [classes/:?]
	at com.cloudant.search3.SearchServer$SearchServerHandler.channelRead0(SearchServer.java:137) [classes/:?]
	at com.cloudant.search3.SearchServer$SearchServerHandler.channelRead0(SearchServer.java:75) [classes/:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext.access$600(AbstractChannelHandlerContext.java:59) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.channel.AbstractChannelHandlerContext$7.run(AbstractChannelHandlerContext.java:368) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-all-4.1.45.Final.jar:4.1.45.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```

**Comment by rnewson** [comment]
that is unrelated to my change, though it's a problem. It means the data changed _type_ for that field (e.g, "true" to 1).

**Comment by tonysun83** [comment]
+1
